subprojects/inject-generics/src/test/java/org/mockito/internal/util/reflection/generic/GenericTypeMatchTest.java

```diff
@@ -0,0 +1,421 @@
+
+        private Integer[] integerArray;
+        private String[] stringArray;
+        private Object[] objectArray;
+
+        @ParameterizedTest
+        @CsvSource({
+            "integerArray  ,integerArray  ,true",
+            "stringArray   ,integerArray  ,false",
+            "objectArray   ,integerArray  ,false",
+            "stringArray   ,stringArray   ,true",
+            "integerArray  ,stringArray   ,false",
+            "objectArray   ,stringArray   ,false",
+            "objectArray   ,objectArray   ,true",
+            "stringArray   ,objectArray   ,true",
+            "integerArray  ,objectArray   ,true",
+        })
+        public void testArrayTypes(String sourceFieldName, String targetFieldName, boolean matches) throws NoSuchFieldException {
+            Field sourceField = ArrayTypesTest.class.getDeclaredField(sourceFieldName);
+            Field targetField = ArrayTypesTest.class.getDeclaredField(targetFieldName);
+            GenericTypeMatch sourceMatch = GenericTypeMatch.ofField(sourceField);
+            GenericTypeMatch targetMatch = GenericTypeMatch.ofField(targetField);
+            assertEquals(matches, targetMatch.matches(sourceMatch));
+        }
+    }
+
+    static class CollectionBox<T> {
+        private Collection<T> collection;
+    }
+
+    static class SubOfBox<S> extends CollectionBox<S> {}
+
+    static class ConcreteSubOfBox extends CollectionBox<Integer> {}
+
+    static class SubOfSubOfBox<S> extends SubOfBox<S> {}
+
+    static class SubOfConcrete extends ConcreteSubOfBox {}
+
+    static abstract class Change {}
+
+    static class ChangeCollection<T extends Change> {
+        private List<T> changes;
+    }
+
+    @Nested
+    public class NestedFieldTest {
+
+        private CollectionBox<Integer> integerCollectionBox;
+        private CollectionBox<String> stringCollectionBox;
+        private CollectionBox<?> wildcardCollectionBox;
+        @SuppressWarnings("rawtypes")
+        private CollectionBox rawCollectionBox;
+
+        private SubOfBox<Integer> integerSubOfBox;
+        private SubOfBox<String> stringSubOfBox;
+        private SubOfBox<?> wildcardSubOfBox;
+
+        private ConcreteSubOfBox concreteSubOfBox;
+        private SubOfConcrete subOfConcrete;
+
+        private SubOfSubOfBox<Integer> integerSubOfSubOfBox;
+        private SubOfSubOfBox<String> stringSubOfSubOfBox;
+
+        @SuppressWarnings("rawtypes")
+        private ChangeCollection changesTargetRaw;
+
+        private List<Integer> integerList;
+        private List<String> stringList;
+        private Collection<?> wildcardCollection;
+
+        private List<Change> changeList;
+
+        @ParameterizedTest
+        @CsvSource({
+            "integerList         ,integerCollectionBox   ,CollectionBox  ,collection  ,true",
+            "integerList         ,stringCollectionBox    ,CollectionBox  ,collection  ,false",
+            "integerList         ,wildcardCollectionBox  ,CollectionBox  ,collection  ,false",
```

Comment:
In the previous test `stringList` was assignable to `wildcardList`. I would have expected the same here. Why is it false in this case?

Reply:
From these tests, it looks like "CollectionBox" is matched only with the exact generic type match (and it's coherent across all test cases here)
stringList -> wildcardList is a more solid match because it's a one-to-one collection-to-collection match. While a CollectionBox is not a collection by itself. So I guess in the current implementation it considered a reach to match a wildcarded not-itself-a-collection with a specific collection.

Reply:
This was corrected in the work continued in #3389